### PR TITLE
Calculate Docker cache hash precisely from Dockerfile's dependencies

### PR DESCRIPTION
**Pull request by kennytm**
#58549 changed the Docker cache calculation to include every file under `src/ci/docker`, so that when files under `dist-x86_64-linux` is changed, its dependent image `dist-i686-linux` will also be rebuilt.

However, this ultraconservative solution caused the `dist-i686-linux` to be rebuilt every time an irrelevant Dockerfile (e.g. the PowerPC ones) is changed, which increases the building time beyond 3 hours and forcing a spurious but expected failure.

This commit instead parses the Dockerfile itself and look for the actual dependencies. The scripts needs to be copied into the Docker image, which must be done with the COPY command, so we just need to find all lines with a COPY command and add the source file into the hash calculator.

Note: this script only handles single-lined COPY command in the form `COPY src1 src2 src3 dst`, since these are the only variant used inside this repository.

### CODE REVIEW

**Comment by pietroalbini** [comment]
Seems good to me.

@bors r+

**Comment by pietroalbini** [comment]
No, wait, it behaves weirdly when tested locally, let me investigate.

@bors r-

File: src/ci/docker/run.sh [hunk 1]
```diff
@@ -22,7 +22,17 @@ if [ -f "$docker_dir/$image/Dockerfile" ]; then
...
+
+      # Look for all source files involves in the COPY command
+      copied_files=/tmp/.docker-copied-files.txt
+      for i in $(sed -n -e 's/^COPY \(.*\) .*$/\1/p' "$docker_dir/$image/Dockerfile"); do
+        # List the file names
+        find "$docker_dir/$i" -type f >> $copied_files
+      done
+      # Sort the file names and cat the content into the hash key
+      sort $copied_files | xargs cat >> $hash_key
+      cat $copied_files
```
Comment (pietroalbini) [inline src/ci/docker/run.sh]: Is this really needed? Seems like a leftover debugging line.

File: src/ci/docker/run.sh [hunk 2]
```diff
@@ -22,7 +22,17 @@ if [ -f "$docker_dir/$image/Dockerfile" ]; then
       hash_key=/tmp/.docker-hash-key.txt
       rm -f "${hash_key}"
       echo $image >> $hash_key
-      find $docker_dir -type f | sort | xargs cat >> $hash_key
+
+      # Look for all source files involves in the COPY command
+      copied_files=/tmp/.docker-copied-files.txt
```
Comment (pietroalbini) [inline src/ci/docker/run.sh]: Can you `rm -f` the file? This way the hash is consistent even when testing the script locally.

**Review by emilyalbini — CHANGES_REQUESTED** [review]
Other than the two (minor) comments you forgot to include the Dockerfile itself in the hash.

**Pull request merged**